File: src/main/java/com/raoulvdberge/refinedstorage/inventory/item/FilterItemsItemHandler.java
package com.raoulvdberge.refinedstorage.inventory.item;

import com.raoulvdberge.refinedstorage.util.StackUtils;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.util.NonNullList;
import net.minecraftforge.items.ItemStackHandler;

public class FilterItemsItemHandler extends ItemStackHandler {
    private ItemStack stack;

    public FilterItemsItemHandler(ItemStack stack) {
        super(27);

        this.stack = stack;

        if (stack.hasTag()) {
            StackUtils.readItems(this, 0, stack.getTag());
        }
    }

    @Override
    protected void onContentsChanged(int slot) {
        super.onContentsChanged(slot);

        if (!stack.hasTag()) {
            stack.setTag(new CompoundNBT());
        }

        StackUtils.writeItems(this, 0, stack.getTag());
    }

    public NonNullList<ItemStack> getFilteredItems() {
        return stacks;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/item/FilterItemsItemHandler.kt
package com.refinedmods.refinedstorage.inventory.item

import com.refinedmods.refinedstorage.util.StackUtils
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.NonNullList
import net.minecraftforge.items.ItemStackHandler

class FilterItemsItemHandler(private val stack: ItemStack) : ItemStackHandler(27) {
    protected fun onContentsChanged(slot: Int) {
        super.onContentsChanged(slot)
        if (!stack.hasTag()) {
            stack.tag = CompoundTag()
        }
        StackUtils.writeItems(this, 0, stack.tag)
    }

    val filteredItems: NonNullList<ItemStack>
        get() = stacks

    init {
        if (stack.hasTag()) {
            StackUtils.readItems(this, 0, stack.tag)
        }
    }
}